src/common/unitTests/testCArray.cpp

```diff
@@ -265,11 +265,151 @@ void testParenthesesOperatorRT()
 }
 TEST( testCArray, testParenthesesOperator )
 {
+#if !defined( SHIVA_USE_BOUNDS_CHECK ) || !defined(SHIVA_USE_DEVICE)
   testParenthesesOperatorCT();
+#endif
   testParenthesesOperatorRT();
 }
 
 
+template< int DIM >
+void testBoundsCheckParenthesesOperator1dHelper( int const ilower = 0,
+                                                 int const iupper = DIM - 1 )
+{
+  pmpl::genericKernelWrapper( [ilower, iupper] SHIVA_DEVICE ()
+  {
+    CArrayNd< double, DIM > array{ 0.0 };
+    for ( int i = ilower; i <= iupper; ++i )
+    {
+      array( i ) = i;
+    }
+  } );
+}
+TEST( testCArray, testBoundsCheckParenthesesOperator1d )
+{
+  testBoundsCheckParenthesesOperator1dHelper< 2 >();
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator1dHelper< 2 >( -1, -1 );}, "" );
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator1dHelper< 2 >( 2, 2 );}, "" );
+}
+
+template< int DIM0 = 2, int DIM1 = 4 >
+void testBoundsCheckParenthesesOperator2dHelper( int const ilower = 0,
+                                                 int const iupper = DIM0 - 1,
+                                                 int const jlower = 0,
+                                                 int const jupper = DIM1 - 1 )
+{
+  pmpl::genericKernelWrapper( [ilower, iupper, jlower, jupper] SHIVA_DEVICE ()
+  {
+    CArrayNd< double, DIM0, DIM1 > array{ 0.0 };
+    for ( int i = ilower; i <= iupper; ++i )
+    {
+      for ( int j = jlower; j <= jupper; ++j )
+      {
+        array( i, j ) = i * DIM1 + j;
+      }
+    }
+  } );
+}
+TEST( testCArray, testBoundsCheckParenthesesOperator2d )
+{
+  testBoundsCheckParenthesesOperator2dHelper();
+
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator2dHelper( -1, -1, 0, 3 );}, "" );
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator2dHelper( 2, 2, 0, 3 );}, "" );
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator2dHelper( 0, 1, -1, -1 );}, "" );
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator2dHelper( 0, 1, 4, 4 );}, "" );
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator2dHelper( -1, -1, -1, -1 );}, "" );
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator2dHelper( -1, -1, 4, 4 );}, "" );
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator2dHelper( 2, 2, -1, -1 );}, "" );
+  EXPECT_DEATH( {testBoundsCheckParenthesesOperator2dHelper( 2, 2, 4, 4 );}, "" );
+
+}
+
+template< int DIM = 3 >
+void testSquareBracketOperator1DHelper( int const ilower = 0,
+                                        int const iupper = DIM - 1 )
+{
+  pmpl::genericKernelWrapper( [ilower, iupper] SHIVA_DEVICE ()
+  {
+    CArrayNd< double, DIM > array{ 0.0 };
+    for ( int i = ilower; i <= iupper; ++i )
+    {
+      array[i] = i;
```

Comment:
static_assert tests that operator[] is returning a reference type here... it is, but being explicit about the test would be good

Reply:
done